src/components/utils.ts

```diff
@@ -0,0 +1,65 @@
+import { FlatOption, Option } from "./types";
+
+export const flattenOptions = (array: FlatOption[], options: Option[], path: string[] = []) => {
+    options.forEach(op => {
+        const hasChildren = op.hasOwnProperty("children");
+        const newPath = [...path, op.id];
+        const isTopLevel = newPath.length === 1;
+
+        const baseOption = {
+            id: op.id,
+            label: op.label,
+            path: newPath,
+            show: isTopLevel
+        };
+
+        if (hasChildren) {
+            array.push({...baseOption, hasChildren, open: false});
+            flattenOptions(array, op.children!, newPath);
+        } else {
+            array.push({...baseOption, hasChildren});
+        }
+    });
+};
+
+const arraysAreEqual = (array1: string[], array2: string[]) => {
+    return JSON.stringify(array1) === JSON.stringify(array2);
+};
+
+export const expandOptions = (array: FlatOption[], optionPath: string[]): FlatOption[] => {
+    const openOptionsPaths: string[][] = [optionPath];
+    return array.map(op => {
+        if (arraysAreEqual(op.path, optionPath)) {
+            return { ...op, open: true };
+        }
+
+        // if you are a child of a previously opened option then you should show
+        const show = openOptionsPaths.some(openOpPath => {
```

Comment:
I don't really understand why previously opened paths are relevant here. Shouldn't it just be modifying show to true for anything immediately under the selected path, and leaving everything else alone? 
And would it make sense to take advantage of the fact that we know the options are in order, so that if we find the next sibling (or higher?) of the selected path we can stop changing things (and we don't need to change anything before the selected path)?
So, could you have something like this (for modifying in place)?
```
const expandOptions = (array: FlatOption[], optionPath: stirng[]) => {
    let index = array.findIndex((fo: FlatOption) => arraysAreEqual(fo.path, optionPath)) + 1;
    // loop until end of all options, or next sibling (or higher level option)
    while (index < array.length && array[index].path.length > optionPath.length ) {
        // only change immediate children
        if (array[index].path.length === optionPath.length + 1) {
            array[index+1].show = true;
        }
        index = index + 1;
    }
}
```
Alternatively while loop could check that the path starts with the optionPath path (rather than just looking at path length).

Reply:
i really like this idea of just mutating that slice of the array, vue seems to work well with just the ref so all good on that front too! the open paths are relevant because they keep track of what someone has opened before and we should open that. For example lets say something expands the options "parent" > "child" > "grandchild" and then they collapse parent. When they expand parent they should see grandchild too, this is how the current treeselect does it and I think it is actually quite a helpful functionality especially in deeper structures

Reply:
Ok, and that's going to use the 'open' prop which is retained rather than 'show' which just determines whether it's being displayed right now? Don't think I'd quite spotted the distinction there, but yeah that makes total sense.